### ISSUE TITLE
Allow revocation list for client cert auth

### ISSUE THREAD

**Issue by Alex131089** [mholt/caddy#2341]
Hello, could caddy add an option to check client certificate against a revocation list ?
This was the object of https://github.com/mholt/caddy/issues/1904, which it was closed as duplicate of https://github.com/mholt/caddy/issues/1375, but https://github.com/mholt/caddy/issues/1375 doesn't seems to implement that.
Best,

### 1. What version of Caddy are you using (`caddy -version`)?
v0.11.0

### 2. What are you trying to do?
Disallow revoked client certs

### 3. What is your entire Caddyfile?
*Not relevant, there's no option to do that as per https://caddyserver.com/docs/tls*

**Comment by shynome** [comment]
need this feature +1.
I like the caddy simple config but I have to go back nginx because nginx support and this is required for security when use tls client certificate.
At last, thank you for so great software

**Comment by allardmuis** [comment]
@mholt 
I'd really like to have this feature as well. I'll try to describe my usecase and why other solutions are not working for me.

I have a system where I have to grand access to my api to other servers. Which other parties are allowed is decided by a government institute. Anyone can apply, and if successfully audited, gets a certificate signed by their CA. So there are many certificates out there, I don't know which ones, new ones can be added every day.
So I allow any client-certificate signed by the CA, without knowing in advance which specific certificates there are. This is easy with Caddy: i just set the CA in 'clients' option.

If a certificate is compromised, or someone fails the yearly audit, the CA revokes the certificates of that party. The certificate is put on the CRL maintained by the CA. I have to check that, otherwise I keep granting access to systems that should not have it anymore.
Even better would be to use OSCP and request to the CA if a client-certificate that is used is still valid.

TLDR; I think CRLs are the only way if you don't hand out the client-certificates yourself, but trust everything signed by an external CA.

Btw: Reading CRLs from file would be much nicer than having them in the config file itself: a CRL might contain hundreds of certificates so the config would get bloated quite a bit.

**Comment by Gr33nbl00d** [comment]
I need this feature too, we have exactly the same use case like @allardmuis 
As this is quite urgent for us we need to find some solution soon.
Could be that im forced to do the changes myself and open a pull requests. But as i have nearly no expirience with golang this will be last solution i would like to go.

@mholt 
you might have a look here as example:
https://github.com/cloudflare/cfssl/blob/master/revoke/revoke.go

**Comment by mholt** [comment]
@Gr33nbl00d 

> I need this feature too, we have exactly the same use case like @allardmuis
As this is quite urgent for us we need to find some solution soon.
Could be that im forced to do the changes myself and open a pull requests. But as i have nearly no expirience with golang this will be last solution i would like to go.

Absolutely understand -- I can get authorized to work on this feature right away if your company can fund its development. If you fill out this form, we can get the ball rolling today: https://www.ardanlabs.com/my/contact-us?dd=caddy

Thanks for the link, too, that might come in handy!

**Comment by Gr33nbl00d** [comment]
@mholt 
I mentioned it to our product manager today ;) Because of some other problems this is postponed to begin of next year. So we life first without CRL till begin of next year. Than we will see ;)

**Comment by mholt** [comment]
As I said above, we can prioritize this feature if your company can pay for its development; or anyone else is welcome to propose and implement it voluntarily. It's on the list (as the issue is still open) but on the backlog as I work through more pressing priorities.

**Comment by wille-io** [comment]
Sorry, didn't want to exert pressure or anything, just wanted to state my point of view.
I wish I knew Go, I can't even find where the variable "ClientCAs" is used, so I could see where to put the counter part for it and implement it myself for Caddy.

**Comment by mholt** [comment]
@wille-io 

>  I can't even find where the variable "ClientCAs" is used,

It's used by the Go standard library's `crypto/tls` package. You can find it used in there.

**Comment by Gr33nbl00d** [comment]
Hi again, nearly 2 years later and we are now in heavy need to implement this feature.
I am currently trying out how to solve things. I was thinking about using OCSP instead of CRL but:
But it seems that you only support OCSP for server certificates not for client certificates.
It seems that OCSP and CRL checking needs to be manually implemented via verifyPeerCertificates function.

Is that correct? If so i can now decide to implement CRL or OCSP support for client certificates or maybe even both ;)

**Comment by mholt** [comment]
The Go standard library will first perform its own verification of the certificate, so if it specifies an OCSP server I imagine it would take care of that for us, yeah?

We can still make a CRL configurable. You want to have a go at it?

**Comment by Gr33nbl00d** [comment]
i just debugged the tls handshake in crypto/tls/handshake_server.go
and it does not look like if there is any OCSP check regarding the client certificate.

It looks like you have to implement checking of client certificates yourself by implementing it in verifyPeerCertificate method.

I found some other go project "snowflakedb" who seemed to have similar issue:

https://github.com/snowflakedb/gosnowflake/issues/5

They have implemented it basically here:
https://github.com/snowflakedb/gosnowflake/blob/master/ocsp.go

A very basic solution for ocsp checks with the existing ocsp functions can be found here:
https://stackoverflow.com/questions/46626963/golang-sending-ocsp-request-returns

Ofcourse there is caching and a lot of important stuff missing in the last example.

**Comment by Gr33nbl00d** [comment]
Also interesting is that issue: github.com/golang/go/issues/36736
Ok im out for today ;) Need some sleep and think about further steps tomorrow ;)

**Comment by Gr33nbl00d** [comment]
Regarding your question if i give it a go. Yes i will start with this. We need that feature very soon. Currently i am just checking the possibilities. Im on vacation the next 3 weeks so i have only the next 2 days to collect some info to get a decisions from management if i should go for crl or ocsp or for both. Ocsp might require enrollment of new client certs on our customers as they are using crls atm and im not sure if this is possible. But i think ocsp is the better way. So maybe we need both in the end ;)

**Comment by mholt** [comment]
Hm, indeed, OCSP verification hasn't been implemented yet, but it has been proposed, and that is what I was thinking of: https://github.com/golang/go/issues/40017

These are rather complex (as Filippo rightly points out) and I hesitate to implement that ourselves into Caddy.

I do think OCSP is probably better, but both have their uses in some cases I believe. If we do implement it ourselves, I'd rather switch to using the Go standard library when they do finally implement it. So whoever contributes this to Caddy will need to be willing to replace it with the standard lib's implementation if/when it becomes available.

**Comment by Gr33nbl00d** [comment]
For sure every line which can be deleted is a good line of code. However i digged a bit deeper into the parsing of crl. One thing i think was also a reason why they did not want to implement crl is that the asn1 implementation of go is more "dom" based. That means reading a whole asn1 structure in memory. And that's always a bad idea if you need to parse indefinitely growing structures like crl. I am about to implement a stream based asn1 parser right now as there is none out there. So you can work with every size of crl. Its a bit childish of the go developers to try to simply ignore that fact. Yeah crl are something that might go away in 10 years but still a lot of good working software depends on it. Also companys depend on it. A fact you should not ignore. To implement both things is not an easy task luckily its not so hard for me ;) i was even thinking to make this available as library so others can benefit but for the first shot i will keep it in caddy. As this problem is open for so long i dont expect that a solution will arrive soon.

**Comment by mholt** [comment]
Yeah, I do not want a separate/new parser in Caddy (or as a dependency,  either).

**Comment by Gr33nbl00d** [comment]
@FiloSottile @mholt  your meantioned parsers are all memory based parsers not stream based parsers.
This is not a good idea if you need to parse data with undefined size. What will happen if you get a 8GB revocation list?
Thats why i am implementing a stream based parser does not mean i parse everything manual ofcourse i use encoding/asn1 and crypto stuff. I simply do not parse the asn1 sequences with all revocation entrys at once but i parse them one by one.
Just have a litle bit of trust that i know what i am doing ;) You can review my code an than you can complain if you like.
This is ofcourse only needed for CRL parsing. OCSP is in way easier to implement.

**Comment by Gr33nbl00d** [comment]
> Not sure what is being discussed here. Is it implementing CRL logic, an OCSP verifier, or a new ASN.1 parser?
> 
> Checking CRLs should be relatively easy, although you'd have to make decisions about the frequency at which to fetch them, and take care to extract the location of the CRLs properly. This is the kind of things you want to ask Sleevi about.
> 
> An OCSP verifier is an extremely nuanced component. See [golang/go#40017](https://github.com/golang/go/issues/40017).
> 
> I do not understand the purpose of a new ASN.1 parser. golang.org/x/crypto/cryptobyte or even encoding/asn1 are perfectly fine for CRL or OCSP support. It would be weird if Caddy grew and had to maintain a whole ASN.1 parser, and then adapt to all X.509 quirks, just to parse CRLs.

Just google for Adam Langley and revocation lists there is a huge discussion. He simply thinks that revocation checks do not make sense CRL based. But a lot of security experts see this different. And the industry still relies a lot on CRL. Even OCSP is often backed by CRL.
Adam Langley has a quite ignorant meaning on that Topic. I work for a big company and our customers want CRL no matter what Adam Langley says

**Comment by Gr33nbl00d** [comment]
I can also implement it as a seperate module if you make it possible for modules to implement the verifyPeerCertificate or in general client certificate checking. This way the whole stuff could be outside of caddy.

**Comment by francislavoie** [comment]
> I can also implement it as a seperate module if you make it possible for modules to implement the verifyPeerCertificate or in general client certificate checking. This way the whole stuff could be outside of caddy.

Feel free to open a PR for this. It'll be easier to discuss with a proof-of-concept implementation.

**Comment by Gr33nbl00d** [comment]
Ok i will. Like i wrote earlier im atm on vacation. I already finished parsing of crl for v1 and v2 two weeks ago. But due vacation i will continue on mid September :)

**Comment by Gr33nbl00d** [comment]
So today i finished signature validation of CRL. I supported all kind of ECDSA and RSA signatures based on OID definition in CRL. But i still have some points where i dont know how it is best solved.

My idea is to support multiple list of CRL locations in the configuration directly, but also support dynamic adding of CRLs via CRL Distribution Points (CDP) Extension in certificates. The question here is if dynamic CRL locations are added via CDP entry in the certificate i somehow need to check if we can trust the certificate from the server handing out the CRL. For that i need some kind of list of trusted root certificates which would authorize the ca certs from the CDP points. Do you think its enough to maintain a list of certificate files in the configuration which can be trusted? This is like a "truststore"
Could look like this trusted_root_certificates {root1.pem root2.pem] or is there already some kind of solution for trusted root certs in caddy.

Also i think its a good idea to allow overriding of CDP locations as sometimes host names/IPs change even if the certificate is already handed out and still valid. For that i would like to implement something like cdp_location_overrides in configuration where you could override a location of a CRL defined inside the certificate so you can adopt the config if the IP/Host changes without changing all certificates handed out. It would be nice to here some feedback of users needing that feature to get a better understanding of the use cases. Otherwise i will implement it like described.


When finished i will work on the extraction into a seperate module by adding some possibilty to implement verifyClientCertificate inside a module

**Comment by francislavoie** [comment]
> Could look like this trusted_root_certificates {root1.pem root2.pem] or is there already some kind of solution for trusted root certs in caddy.

Well, there's [`acme_ca_root`](https://caddyserver.com/docs/caddyfile/options#acme-ca-root) but that's just for the ACME CA, not for all usecases. There's also [`tls_trusted_ca_certs`](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#the-http-transport), but that's just for the `reverse_proxy` handler, since it's a client to the upstream app.

I figure CRL stuff would be configured somewhere in the [`tls` app](https://caddyserver.com/docs/json/apps/tls/) and the trusted certs config would look largely similar to the aforementioned config options, but just for the CRL lookups.

**Comment by Gr33nbl00d** [comment]
> > Could look like this trusted_root_certificates {root1.pem root2.pem] or is there already some kind of solution for trusted root certs in caddy.
> 
> Well, there's [`acme_ca_root`](https://caddyserver.com/docs/caddyfile/options#acme-ca-root) but that's just for the ACME CA, not for all usecases. There's also [`tls_trusted_ca_certs`](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#the-http-transport), but that's just for the `reverse_proxy` handler, since it's a client to the upstream app.
> 
> I figure CRL stuff would be configured somewhere in the [`tls` app](https://caddyserver.com/docs/json/apps/tls/) and the trusted certs config would look largely similar to the aforementioned config options, but just for the CRL lookups.

I will try to do it in a similar or equal way than the other solutions are done

**Comment by Gr33nbl00d** [comment]
For those who are interested why i had to do some extra work on the asn1 parsing:
I did some final tests and measurements regarding CRL size and memory usage as well as performance tests.
For testing i used a real large CRL from the department of defense (Email CA-59)
(the largest i could find)

https://crl.gds.disa.mil/getcrlzip?DOD+EMAIL+CA-59

This CRL is about 48 MB

By parsing it with standard go way in a main method (Unmarshalling the whole list in one big chunk)
```
crlBytes, err := ioutil.ReadFile("DODEMAILCA_59.crl")
crl, err := x509.ParseCRL(crlBytes)

```

Result is following:
After first line memory usage is about 50 MB thats simply the whole file into memory.
After parsing (second line) memory usage grow to 423MB

My stream based version which uses partial unmarshalling does not exceed 8MB memory usage while parsing.
So i am quite happy with the result.

Parsing times are quite equal 3.9 seconds versus 4.2 seconds with the stream based approach.

**Comment by Gr33nbl00d** [comment]
I am big steps further i did a first successfull CRL check today, but there is still a lot todo. 
There are some decisions to make and some questions i have you might can answer.
Let me know what you think

For information i will first try to do a MVP release (phase 1) (minimum viable product) which cover most use cases.
I will then add the other more rare extradorinary use cases in a second step (phase 2)

### Question i could not find an answer to yet
The verifyPeerCertificate method has an array of certificate chains (array of array of certificates)
`verifiedChains [][]*x509.Certificate`
I was wondering under which circumstances there can be multiple chains. I would have expected only one chain.
Can someone answer this question?

### Design proposal and phases
* The CRLDistributionPoints can contain multiple schemes to download the crl (HTTP,LDAP...)
  Phase 1:  Only support HTTP not LDAP,FTP... If no HTTP distribution point is found fail the connection
  Phase 2: Consider to add LDAP or other schemes (Allow to configure a preference (prefer HTTP or prefer LDAP)

* What happens if the CRL signature can not be validated? 
   Phase 1: Make the connection fail
   Phase 2: Make it configurable to disable CRL signature checking

* What happens if the CRL defined in CRLDistributionPoints can not be downloaded at all or in a specific time
   Phase 1: Make the connection fail
   Phase 2: Make it configurable to fail or ignore (might allow some 'grace period' epecially for the first connection attempt)

* What happens if the CRL is outdated possible because no update could be downloaded due server downtime
   Phase 1: Ignore this circumstance
   Phase 2: Make it configurable to fail or ignore (might allow some 'grace period')

* The CRLDistributionPoints inside the certificate can be outdated because of server movement url changes 
   Phase 2: Allow overriding CRLDP locations via config

* According to rfc5280 section 4.2.1.13. The CRL issuer does not need to be the certificate issuer. In most cases the CRL issuer is the certificate issuer. In that case we can be sure that we have the certificate necessary for CRL signture validation (inside the chain). In the other case we need to maintain a list of trusted CRL issuer certificates in some kind of truststore
  Phase 1: Assume CRL issuer is the cert issuer or at least part of the certificate chain. If the certificate can not be found in the chain fail the connection
  Phase 2: Allow to configure a list of trusted CRL issuer certificates

**Comment by mholt** [comment]
> I was wondering under which circumstances there can be multiple chains. I would have expected only one chain.
> Can someone answer this question?

A cert bundle is just a pile of certificates, and there may be varying ways a chain of certs can be formed, depending on the algorithm. I'm not 100% sure what Go does without reading into it.

**Comment by Gr33nbl00d** [comment]
Thank you for your answer. Luckily i found the reason myself in a quite long article. I think i know now what to do with these multiple chains.
https://utcc.utoronto.ca/~cks/space/blog/tech/TLSHowMultipleChains

**Comment by francislavoie** [comment]
Commenting on the config, comma separated string values is pretty awkward, I recommend using string slices for that instead. For example, `"mode": "prefer_ocsp,prefer_crl"` should probably be `"mode": ["prefer_ocsp", "prefer_crl"]` instead.

Since it's a pretty niche need to revocation, I think I'd rather this be an external module. I don't think @mholt or I would want to inherit maintenance of this functionality either, if it's merged into this repo. Doesn't feel good to merge code that we don't completely understand and feel we can fix bugs in, especially for security sensitive code.

What I suggest is that a "generic" API for this is designed such that we can add an extension point to the `tls` module to implement revocation strategies, or something like that. I.e. a new module namespace to cover this: https://caddyserver.com/docs/extending-caddy/namespaces. I'm not totally sure "when" revocation checks are triggered, so I'm not sure how that should look, but hopefully you have some ideas.

**Comment by francislavoie** [comment]
Okay, interesting. Maybe we could refactor client auth modes to be implemented as a module namespace, each mode being a module that implements that interface.

The existing modes could remain, and you could add your own like `require_and_verify_with_revocation` or something like that.

The shape of the config might need to change, but I think that's probably fine as long as the same behaviour remains.

**Comment by Gr33nbl00d** [comment]
Hi again :) One day later and i finally think i now have a bit better understanding.
Ok one easy way to solve my problems is by adding this to the client_authentication

```
{
  "client_authentication": {
    "mode": "require_and_verify",
    "trusted_ca_certs_pem_files": [
      "./certificates/ca.crt"
    ],
    "client_cert_validations": {
      "revocation_list": {
      }
    }
  }
}
```

In the ClientAuthentication i would add following fields:
```
	RawClientCertValidations RawClientCertValidations `json:"client_cert_validations,omitempty" caddy:"namespace=tls.clientcertvalidations"`
	ClientCertValidations []ClientCertValidator `json:"-"`

```

This way additional modules providing a way to verify client certs can be added. Nothing else need to be changed.
However if you look at the config it makes things a bit inconsistent. As the mode also defines "validations"

I think the real problem is inside go crypto funny that this was also done by adam langley ;) 
"mode" has some consistency problem it defines actually two things, request mode and validation mode
cert_request_mode=no_request,request 
cert_validation_mode=no_validation,validation

If it was not this way it could have been implemented like this:

{
  "client_authentication": {
    "mode": "request",  --> defines if to request or not to request a client cert
    "client_cert_validations": {
      "require": {},                  ---> If configured checks if a cert is given and if not fails TLS handshake
      "validate_signature": {  ---> If client cert is present validate cert
        "trusted_ca_certs_pem_files": [
          "./certificates/ca.crt"
        ]
      },
      "revocation_list": {     ---> If client cert is present check crl
      }
    }
  }
}

But as this is a problem inside Go crypto itself not valueing SRP (Single Responsibility Principle) there is no easy way without completly changing the handshake algorithm. Only way would be that you configure only the "request mode" in go tls config this means only NoCLientCert or RequestClientCert. And do the validation (require cert/validate cert) yourself in different implementations of verifyPeerCertificate implementing ClientCertValidator

If you are ok i will prepare a pull request implementing it like described in first example. It is still possible to do changes like described below afterwards.

**Comment by Gr33nbl00d** [comment]
I am always for good naming. The names above were not final just for example. So yeah i will do it like this :)

**Comment by Gr33nbl00d** [comment]
I have a question regarding cyclic updates. I need to do cyclic updates of CRL Lists. Therefore i need some kind of time based trigger. I did this in the past by adding code to caddy tls modules "start" function, creating another ticker for my use case.

As this module is now independent i wonder how to solve this.
I see there multiple ways:

* I could either define an own application to have my own start function
Drawback is that you would need to configure the application in the caddy config + the validator in the tls section

* I could start the ticker in the provision method of my module --> i think this is unclean also no stop method is possible

* Another way would be to add another namespace for cyclic processings --> could be complicated as multiple tickers would need to be started with possible different intervals.  Lifecycle of these tickers need to be maintained.

Mayber there are even other possibilities? What would you suggest?

**Comment by Gr33nbl00d** [comment]
In the end i changed my mind and used option 2 (provision and cleanup)

However first version is commited. Still some minor features missing.


https://github.com/Gr33nbl00d/caddy-revocation-validator

**Comment by francislavoie** [comment]
@christoph-kluge mainly because it's a large feature that we don't want the maintenance burden for. We need to be careful with the code we add to the codebase. Also, it's not likely to be a feature the general public will need, and those that need it should be capable of making their own builds of Caddy with that feature added (it's very easy with `xcaddy`)

**Comment by christoph-kluge** [comment]
How do you think about removing the client-certificate feature from the core and move it into an external module? This would allow custom builds/extensions/forks of it.

I have the feeling that this would be a clean separation instead of having part of the functionality inside the core.

**Comment by evilmonkey19** [comment]
Hi! I just want to add to this conversation that let's encrypt is deprecating OCSP. I don't know if Caddy internally for achieving the automatic TLS use this protocol. They recommend to use CRLs.

This is the link https://letsencrypt.org/2024/07/23/replacing-ocsp-with-crls.html

**Comment by Gr33nbl00d** [comment]
> Hi! I just want to add to this conversation that let's encrypt is deprecating OCSP. I don't know if Caddy internally for achieving the automatic TLS use this protocol. They recommend to use CRLs.
> 
> This is the link https://letsencrypt.org/2024/07/23/replacing-ocsp-with-crls.html

Interesting. As far as i know ocsp is used for serverside certificates by caddy. This issue here was only about client certificates. But there is a problem as golang does not really support crls. The main issue is that they dont support asn1 stream based parsing. Because of this you can only load the whole crl in memory which is potentially problematic because crl size depends on crl entries. The crypto stack was developed by adam langley a google developer and security specialist who has a very strong opinion on revocation... Have a look here:
https://stackoverflow.com/questions/37058322/how-can-i-verify-client-certificates-against-a-crl-in-golang 

Btw this opinion is not shared by all specialists. Because of this crls are not really supported and golang officially only supports ocsp as they considered crls as outdated anyway. I have solved this for the client certs at least but had to write my own asn1 parsing from scratch

**Comment by mholt** [comment]
Welp. That's a disappointing development.

@Gr33nbl00d You have a streaming CRL parser right? Is there a Go package with godocs that we could look into using? I think that may be the only reasonable way if we are to persist Caddy's famous auto-replace-on-revocation abilities.

**Comment by Gr33nbl00d** [comment]
> Welp. That's a disappointing development.
> 
> @Gr33nbl00d You have a streaming CRL parser right? Is there a Go package with godocs that we could look into using? I think that may be the only reasonable way if we are to persist Caddy's famous auto-replace-on-revocation abilities.

Yes i have a stream based parser but jo godocs. It was hard to implement as CRLs are really complicated and i totally understand that someone struggles or refuses to implement this.  There was nothing out there on the net. I checked how it was implemented in java while implementing it. But just parsing is not enough that alone is complicated because of various crl extensions you might need to consider during parsing. You somehow need also to store the crl entries in a way that you can access/search them quickly. I have done this via leveldb or alternatively a map in memory. I have not written any godoc simply because of time saving and not expecting someone will need it as it is not a library. However i could think of extracting it into a library. Would be maybe useful. . the parsing is quite simple done: You can see it here in this test: 


https://github.com/Gr33nbl00d/caddy-revocation-validator/blob/main/crl/crlreader/crlreader_test.go
However you also need to store the data somewhere thats actually done by the crlpersisterprocessor. I mean you can feel free to take my code to use it but the bad thing is tha the topic is complex :( Because crls are complex. I guess it wasnt the answer that you hoped for. Maybe golang devs change there mind and will implement support by this development ?

**Comment by mholt** [comment]
@Gr33nbl00d How interested would you be in helping me integrate your parser in a branch of CertMagic?

**Comment by Gr33nbl00d** [comment]
I am interested, but also have to say i am currently a little busy at work and will be out of office from 26.8 till 13.9. So depends on your expectation. I was already thinking how that could be done. In which intervals do you check currently? I guess the use case regarding server certs is a bit less complex as the certs do not change often and we talk about a limited number of certs. Could be that some parts of the code could be left out because of this. Maybe we can di it that way that we integrate it in a quick way without much changes and you later refactor it to the needs for example remove some unneeded parts. Also depends on the time line. Letsencrypt wrote 3 to 6 month but they waiting for some adoption by Microsoft which should be out in 6 to 12 months. So my guess would be not before 6 months. More likely 12 months but we are not sure. Btw. Traefik golang based load balancer and used in some kubernetes implementations will have the same issue.

**Comment by mholt** [comment]
Sorry for the late reply.

We currently scan for OCSP needing refresh every hour, and renew halfway through OCSP lifetime. We'll need all new logic for CRLs if we go that route, I imagine.

I'm not in a rush to implement this... the goal would be to scan CRLs as often as needed to be able to detect revocations and automatically replace the certs. I suppose we'd need a list of CRLs (in addition to the CA CRLs linked in the certs, I guess) and a streaming CRL parser. It's too bad CRLs don't have a O(1) lookup mechanism.

**Comment by Gr33nbl00d** [comment]
Yes you are right. For client certs i need to keep the crls fast accessible because i need to scan a lot of different certs on each connection. So i currently cash them in a leveldb or in a memory map which is updated in regular intervals. For server side checking of certificates i guess it would be enough to download crl and parse through it stream based. While parsing checking the if the certificate is inside. The good thing is we dont need to remember anything as wen can forget about it once finished with parsing. So we dont need to worry about memory usage. Maybe i could extract some of the coding with some changes to better seperate the download and parsing from the other parts currently implemented. Than it could be easy to add to caddy.

**Comment by Gr33nbl00d** [comment]
I am currently quite busy with some project which i need to finish before vacation. So i guess before mid/end of september (after my vacation) i wont be able to do anything :(
